frame/executive/src/lib.rs

```diff
@@ -287,10 +287,15 @@ where
 	///
 	/// This should only be used for testing.
 	pub fn try_runtime_upgrade() -> Result<frame_support::weights::Weight, &'static str> {
-		<(COnRuntimeUpgrade, AllPalletsWithSystem) as OnRuntimeUpgrade>::pre_upgrade().unwrap();
+		// ensure both `pre_upgrade` and `post_upgrade` won't changing the storage root
+		let state = {
+			let _guard = frame_support::StorageNoopGuard::default();
```

Comment:
Why do we need this? Not sure we should do this.

Reply:
When `StorageNoopGuard` drops it will check the storage root is the same as the one when it is created, thus ensuring the storage root is not changed during `pre_upgrade`.

Reply:
Ensuring the storage root not changing is also one of the targets of this PR, see https://github.com/paritytech/substrate/issues/10824#issuecomment-1105405910

Reply:
hmm okay